File: Python/Algorithm/Tree_Terminology.py
# Complexity: O(n)

tre = [[] for _ in range(N)]
par = [-1] * N; lvl = [0] * N; sz = [0] * N
def dfs(u, p=-1):
    par[u] = p; sz[u] = 1
    for v in tre[u]:
        if v != p:
            lvl[v] = lvl[u] + 1
            dfs(v, u); sz[u] += sz[v]

# Example operation to read a tree and perform DFS
eg = int(input())
for _ in range(eg):
    u, v = map(int, input().split())
    tre[u].append(v); tre[v].append(u)
dfs(1)
for i in range(1, eg + 2):
    print(f"Vertex {i}: Parent = {par[i]}, Level = {lvl[i]}, Subtree Size = {sz[i]}")

File: CPP/Algorithm/Tree_Terminology.cpp
// Complexity: O(n)

vector<long long> tre[N]; long long par[N], lvl[N], streSz[N];
auto dfs(auto u, auto p) {
    par[u] = p; streSz[u] = 1;
    for (auto &v : tre[u]) {
        if (v != p) {
            lvl[v] = lvl[u] + 1;
            dfs(v, u); streSz[u] += streSz[v];
        }
    }
}

// Operation:
    int eg; cin >> eg;
    for (i = 0; i < eg - 1; ++i) {
        int u, v; cin >> u >> v;
        tre[u].push_back(v); tre[v].push_back(u);
    }
    lvl[1] = 0; dfs(1, -1);
    for (i = 1; i <= eg; ++i) {
        cout << i << ' ' << par[i] << ' ' << lvl[i] << ' ' << streSz[i] << '\n';
    }